contracts/src/Consensus.sol

```diff
@@ -3,163 +3,183 @@ pragma solidity ^0.8.30;
-        _requireValidRollover(epochs, proposedEpoch, rolloverAt);
+    function proposeEpoch(uint64 proposedEpoch, uint64 rolloverBlock, FROSTGroupId.T group) public {
+        Epochs memory epochs = _processRollover();
+        _requireValidRollover(epochs, proposedEpoch, rolloverBlock);
         Secp256k1.Point memory groupKey = _COORDINATOR.groupKey(group);
-        bytes32 message = epochRolloverMessage(epochs.active, proposedEpoch, rolloverAt, groupKey);
-        emit EpochProposed(epochs.active, proposedEpoch, rolloverAt, groupKey);
-        _COORDINATOR.sign(activeGroup, message);
+        bytes32 message = domainSeparator().epochRollover(epochs.active, proposedEpoch, rolloverBlock, groupKey);
+        emit EpochProposed(epochs.active, proposedEpoch, rolloverBlock, groupKey);
+        _COORDINATOR.sign($groups[epochs.active], message);
     }
 
     /// @notice Stages an epoch to automatically rollover.
-    function stageEpoch(uint64 proposedEpoch, uint64 rolloverAt, FROSTGroupId.T group, FROSTSignatureId.T signature)
+    function stageEpoch(uint64 proposedEpoch, uint64 rolloverBlock, FROSTGroupId.T group, FROSTSignatureId.T signature)
         public
     {
-        (Epochs memory epochs, FROSTGroupId.T activeGroup) = _processRollover();
-        _requireValidRollover(epochs, proposedEpoch, rolloverAt);
+        Epochs memory epochs = _processRollover();
+        _requireValidRollover(epochs, proposedEpoch, rolloverBlock);
         Secp256k1.Point memory groupKey = _COORDINATOR.groupKey(group);
-        bytes32 message = epochRolloverMessage(epochs.active, proposedEpoch, rolloverAt, groupKey);
-        _COORDINATOR.signatureVerify(signature, activeGroup, message);
-        $epochs = Epochs({active: epochs.active, staged: proposedEpoch, rolloverAt: rolloverAt, _padding: 0});
-        $groups.staged = group;
-        emit EpochStaged(epochs.active, proposedEpoch, rolloverAt, groupKey);
+        bytes32 message = domainSeparator().epochRollover(epochs.active, proposedEpoch, rolloverBlock, groupKey);
+        _COORDINATOR.signatureVerify(signature, $groups[epochs.active], message);
+        epochs.staged = proposedEpoch;
+        epochs.rolloverBlock = rolloverBlock;
+        $epochs = epochs;
+        $groups[proposedEpoch] = group;
+        emit EpochStaged(epochs.active, proposedEpoch, rolloverBlock, groupKey);
     }
 
-    function onKeyGenCompleted(FROSTGroupId.T group, bytes calldata context) external {
-        (uint64 proposedEpoch, uint64 rolloverAt) = abi.decode(context, (uint64, uint64));
-        proposeEpoch(proposedEpoch, rolloverAt, group);
+    function proposeTransaction(MetaTransaction.T memory transaction) external returns (bytes32 message) {
+        Epochs memory epochs = _processRollover();
+        message = domainSeparator().transactionProposal(epochs.active, transaction.hash());
+        emit TransactionProposed(message, epochs.active, transaction);
```

Comment:
Why not just emit the safeTxHash instead of the whole tx. If we index that then it should be easy to find the info for a specific safeTx.

Reply:
`message` is indexed and is kind of the "safe transaction hash" for proposed transactions.

Do you mean additionally emit and index the `transaction.hash()` value?

Reply:
Added - as a side note, we should probably go through all the events at the end and decide "what fields we need and which ones need to be indexed" once we are up and running. I'm 100% OK with erring on the side of caution for now, but I suspect that we are logging some superfluous data that we can remove. Same goes for storage.